### DISCUSSION TITLE
[Chatbot] List out necessary endpoints

### DISCUSSION THREAD

**Discussion by tuantran0910** [TinhHoaSolutions-EzHR/chatbot#122] [General]
- Currently, Frontend team and Backend team are working in this project independently. In the next stage, we are going to integrating works between two teams to make the app run for the MVP.
- This discussion is about defining the necessary endpoints, also centralize them for working.

**Comment by tuantran0910** [comment]
## `GET /settings`
- ~~Get system settings~~ -> Get user and system settings.
- Backend endpoint: `/api/v1/users/me/settings`
- Command:

```bash
curl -X GET 127.0.0.1:5000/api/v1/users/me/settings
```

- Output:

```bash
{
    "message": "Success",
    "headers": null,
    "data": {
        "recent_agent_ids": [
            "c262ab96-06a5-46eb-8120-3cf978d11180",
            "c262ab96-06a5-46eb-8120-3cf978d11176",
            "c262ab96-06a5-46eb-8120-3cf978d11172"
        ],
        "auto_scroll": false,
        "default_model": "o1-pro",
        "maximum_chat_retention_days": null
    }
}
```

**Comment by tuantran0910** [comment]
## `GET /folders`
- Get user's chat folders.
- Backend endpoints: `/api/v1/folders`
- Command:

```bash
curl -X GET 127.0.0.1:5000/api/v1/folders
```

- Outputs:

```bash
{
    "message": "Success",
    "headers": null,
    "data": [
        {
            "id": "b1803c7f-4493-467a-9198-1c8f33d14579",
            "user_id": "f6f7b43c-c0ca-4003-8143-7c5e767cde12",
            "name": "Nhan boi q",
            "display_priority": 0,
            "chat_sessions": []
        }
    ]
}
```

## `POST /folders`
- Create chat folder.
- Backend endpoint: `/api/v1/folders`
- Command:

```bash
curl -X POST 127.0.0.1:5000/api/v1/folders -d '{"name": "Nhan boi q"}'
```

- Output:

```bash
{
    "message": "Success",
    "headers": null,
    "data": {
        "name": "Nhan boi q"
    }
}
```

## `PATCH /folders/{folder-id}`
- Create chat folder.
- Backend endpoint: `/api/v1/folders/{folder-id}`
- Command:

```bash
curl -X PATCH 127.0.0.1:5000/api/v1/folders/<folder-id> -d '{"name": "Thang boi q"}'
```

- Output:

```bash
{
    "message": "Success",
    "headers": null,
    "data": {
        "name": "Thang boi q"
    }
}
```

## `DELETE /folders`
- Create chat folder.
- Backend endpoint: `/api/v1/folders/{folder-id}`
- Command:

```bash
curl -X DELETE 127.0.0.1:5000/api/v1/folders/<folder-id>
```

**Comment by tuantran0910** [comment]
## `GET /chat-sessions`
- Get user's chat sessions.
- Backend endpoint: `/api/v1/users/chat/chat-sessions`
- Command:

```bash
curl -X GET 127.0.0.1:5000/api/v1/chat/chat-sessions
```

- Output:

```bash
{
    "message": "Success",
    "headers": null,
    "data": [
        {
            "id": "c76a4b96-06a5-46eb-8120-3cf978d11112",
            "description": null,
            "user_id": "f6f7b43c-c0ca-4003-8143-7c5e767cde12",
            "agent_id": null,
            "messages": null,
            "folder_id": null,
            "shared_status": "private",
            "created_at": "2024-12-22T09:33:14.437401Z",
            "updated_at": "2024-12-22T09:33:14.439035Z",
            "deleted_at": null
        }
    ]
}
```

## `GET /chat-sessions/{chat-session-id}`
- Get user's specific chat session by ID.
- Backend endpoint: `/api/v1/users/chat/chat-sessions/{chat-session-id}`
- Command:

```bash
curl -X GET 127.0.0.1:5000/api/v1/chat/chat-sessions/<chat-session-id>
```

- Output:

```bash
{
    {
    "message": "Success",
    "headers": null,
    "data": {
        "id": "c76a4b96-06a5-46eb-8120-3cf978d11112",
        "description": null,
        "user_id": "f6f7b43c-c0ca-4003-8143-7c5e767cde12",
        "agent_id": null,
        "messages": [
            {
                "id": "3c6046ab-445a-461c-ad5c-b6f5a7ee108a",
                "chat_session_id": "c76a4b96-06a5-46eb-8120-3cf978d11112",
                "message": "Hi, I am 9h ho lit tich",
                "message_type": "user",
                "parent_message_id": null,
                "child_message_id": "7c653667-f165-4465-8249-eb438908b855",
                "is_sensitive": false,
                "created_at": "2024-12-22T09:33:26.033097Z",
                "updated_at": "2024-12-22T09:37:11.966929Z"
            },
            {
                "id": "7c653667-f165-4465-8249-eb438908b855",
                "chat_session_id": "c76a4b96-06a5-46eb-8120-3cf978d11112",
                "message": "Hi, I am 9h ho lit tich (Response)",
                "message_type": "assistant",
                "parent_message_id": "3c6046ab-445a-461c-ad5c-b6f5a7ee108a",
                "child_message_id": "519da9aa-b0b3-4f35-9040-c56e3bc76ee0",
                "is_sensitive": false,
                "created_at": "2024-12-22T09:33:26.045456Z",
                "updated_at": "2024-12-22T09:33:29.632052Z"
            }
        ],
        "folder_id": null,
        "shared_status": "private",
        "created_at": "2024-12-22T09:33:14.437401Z",
        "updated_at": "2024-12-22T09:33:14.439035Z",
        "deleted_at": null
    }
}
```

## `POST /chat-sessions`
- Create a new chat session.
- Backend endpoint: `/api/v1/users/chat/chat-sessions
- Command:

```bash
curl -X POST 127.0.0.1:5000/api/v1/chat/chat-sessions -d '{}'
```

- Output:

```bash
{
    {
    "message": "Success",
    "headers": null,
    "data": null
}
```

## `PATCH /chat-sessions`
- Update a chat session.
- Backend endpoint: `/api/v1/users/chat/chat-sessions/{chat-session-id}`
- Command:

```bash
curl -X PATCH 127.0.0.1:5000/api/v1/chat/chat-sessions/<chat-session-id> -d '{"description": "Chatboooot"}'
```

- Output:

```bash
{
    {
    "message": "Success",
    "headers": null,
    "data": {
        "description": "Chatboooot"
    }
}
```

## `DELETE /chat-sessions`
- Delete a chat session.
- Backend endpoint: `/api/v1/users/chat/chat-sessions/{chat-session-id}`
- Command:

```bash
curl -X DELETE 127.0.0.1:5000/api/v1/chat/chat-sessions/<chat-session-id>
```

**Comment by lekiet1214** [comment]
Below are listing of vital APIS, I will provide details later.

---
## Chat functions; these get called first and cannot be missing.
- **GET** `/auth/type`
- **GET** `/me`
- **GET** `/manage/indexing-status`
- **GET** `/manage/document-set`
- **GET** `/chat/get-user-chat-sessions`
- **GET** `/query/valid-tags`
- **GET** `/llm/provider`
- **GET** `/folder`
- **GET** `/input_prompt`


---
### All **GET** endpoints
- **GET** `/threads/{thread_id}/runs/{run_id}`
- **GET** `/threads/{thread_id}/runs`
- **GET** `/threads/{thread_id}/runs/{run_id}/steps`
- **GET** `/{assistant_id}`
- **GET** `/threads/{thread_id}/messages`
- **GET** `/threads/{thread_id}/messages/{message_id}`
- **GET** `/{thread_id}`
- **GET** `/get-user-chat-sessions`
- **GET** `/get-chat-session/{session_id}`
- **GET** `/max-selected-document-tokens`
- **GET** `/file/{file_id}`
- **GET** `/global`
- **GET** `/{category}`
- **GET** `/{category}/download`
- **GET** `/document-size-info`
- **GET** `/chunk-info`
- **GET** `/admin/cc-pair/{cc_pair_id}/index-attempts`
- **GET** `/admin/cc-pair/{cc_pair_id}`
- **GET** `/admin/cc-pair/{cc_pair_id}/last_pruned`
- **GET** `/admin/cc-pair/{cc_pair_id}/sync-permissions`
- **GET** `/admin/cc-pair/{cc_pair_id}/get-docs-sync-status`
- **GET** `/admin/connector/failed-indexing-status`
- **GET** `/admin/connector/indexing-status`
- **GET** `/connector`
- **GET** `/connector/{connector_id}`
- **GET** `/indexing-status`
- **GET** `/admin/credential`
- **GET** `/admin/similar-credentials/{source_type}`
- **GET** `/credential`
- **GET** `/credential/{credential_id}`
- **GET** `/admin/indexing-errors/{index_attempt_id}`
- **GET** `/connector-docs/{cc_pair_id}`
- **GET** `/ingestion`
- **GET** `/get-current-search-settings`
- **GET** `/get-secondary-search-settings`
- **GET** `/get-all-search-settings`
- **GET** `/unstructured-api-key-set`
- **GET** `/health`
- **GET** `/auth/type`
- **GET** `/version`
- **GET** `/admin/doc-boosts`
- **GET** `/admin/genai-api-key/validate`
- **GET** `/manage/users`
- **GET** `/manage/admin/valid-domains`
- **GET** `/users`
- **GET** `/get-user-role`
- **GET** `/me`
- **GET** `/{tool_id}`
- **GET** ``
- **GET** `/document-set`
- **GET** `/document-set-public`
- **GET** `/gpu-status`
- **GET** `/authorize`
- **GET** `/admin/query`
- **GET** `/admin/user`
- **GET** `/admin/chat-sessions`
- **GET** `/admin/chat-session-history`
- **GET** `/admin/chat-session-history/{chat_session_id}`
- **GET** `/admin/query-history-csv`

**Reply by tuantran0910** [reply]
Comment following things:
- `GET /gpu-status`
- `GET /document-set`
- `GET /document-set-public`

**Reply by tuantran0910** [reply]
Comment following things:
- ~~`GET /auth/type`~~
- `GET /manage/indexing-status`
- `GET /manage/document-set`
- `GET /query/valid-tags`
- `GET /input_prompt`

**Reply by tuantran0910** [reply]
Comment stuff related to **Token Rate Limit**.

**Reply by tuantran0910** [reply]
Comment following things:
- ~~`GET /admin/query-history-csv`~~

**Reply by tuantran0910** [reply]
Comment following things:
- `GET /admin/doc-boosts`

**Reply by tuantran0910** [reply]
Comment following things:
- `GET /max-selected-document-tokens`
- `POST file`
- `GET /file/{file_id}`
- Related to **Category**

**Comment by lekiet1214** [comment]
## GET `/health`
This is fairly simple, just tell frontend that backend is running and responsive.
Response
```
'message': 'ok'
```

**Reply by tuantran0910** [reply]
Done !

Backend endpoint: `/health`.

**Comment by tuantran0910** [comment]
## `GET /version`
- Get current API server's version.
- Backend endpoint: `/version`
- Command:

```bash
curl -X GET 127.0.0.1:5000/version
```

- Output:

```bash

```

**Comment by tuantran0910** [comment]
## `GET /connectors`
- Get a list of connectors.
- Backend endpoint: `/api/v1/connectors`
- Command:

```bash
curl -X GET 127.0.0.1:5000/api/v1/connectors
```

- Output:

```bash

```

## `GET /connectors/{connector-id}`
- Get specific connector by ID.
- Backend endpoint: `/api/v1/connectors/{connector-id}`
- Command:

```bash
curl -X GET 127.0.0.1:5000/api/v1/connectors/<connector-id>
```

- Output:

```bash

```

## `POST /connectors`
- Create a new connector.
- Backend endpoint: `/api/v1/connectors`
- Command:

```bash
curl -X POST 127.0.0.1:5000/connectors -d '{"name": "Python", "file_paths": ["/opt/python/bin/python3", "/opt/python/local/bin/python3"]}'
```

- Output:

```bash

```

## `PATCH /connectors`
- Update a connector.
- Backend endpoint: `/api/v1/connectors/{connector-id}`
- Command:

```bash
curl -X PATCH 127.0.0.1:5000/api/v1/connectors/<connector-id> -d '{"name": "Pythoon3"}'
```

- Output:

```bash

```

## `DELETE /connectors`
- Delete a connector.
- Backend endpoint: `/api/v1/connectors/{connector-id}`
- Command:

```bash
curl -X DELETE 127.0.0.1:5000/api/v1/connectors/<connector-id>
```
